client/galaxy/scripts/mvc/tool/tool-form-composite.js

```diff
@@ -319,6 +319,19 @@ define([ 'utils/utils', 'utils/deferred', 'mvc/ui/ui-misc', 'mvc/form/form-view'
             }
         },
 
+        /** Refresh the history after job submission while form is shown */
+        _refreshHistory: function() {
+            var self = this;
+            var history = parent.Galaxy && parent.Galaxy.currHistoryPanel && parent.Galaxy.currHistoryPanel;
+            this._refresh_history && clearTimeout( this._refresh_history );
+            if ( history ) {
+                history.refreshContents();
+                if ( history.model.numOfUnfinishedShownContents() === 0 ) {
+                    this._refresh_history = setTimeout( function() { self._refreshHistory() }, 2000 );
+                }
+            }
+        },
+
         /** Build remaining steps */
         _execute: function() {
```

Comment:
I don't understand. Why are you re-implementing this?:
https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/mvc/history/history-model.js#L195

Just call this down on line 397:
`history.refreshContents();`

While any job associated with it is new, queued, or running, the history will poll and keep polling for new datasets. Are the jobs not in one of those states? Are they not associated with a history_id?